Renaming test file causing test class to be deleted

### Current Behavior

I'm finding that if I write a test with the right class name and wrong filename, and then build+run the tests, and then rename the test to have the write filename, that the test stops running because its class no longer gets created

### Expected Behavior

It would be nice if the test would build and run even if it had previously existed with a different filename

### Context (optional)

I was trying to write tests for Gradle itself and found that my test wasn't running, and I suspect that this might be why.

### Steps to Reproduce

To reproduce, run this script inside a checkout of Gradle itself:

```
#!/bin/bash
set -e

separator="************************************************************************"
testname="$1"
if [ "$testname" == "" ]; then
  testname="SampleTest"
fi

function echoAndDo() {
  echo "$*"
  eval "$*"
}

function writeTest() {
  testname="$1"
  filepath="$2"

  echo "$separator"
  echo "Writing failing test at $filepath"

  echo " 
  package org.gradle.internal.execution.history.changes
  
  import spock.lang.Specification
  
  class $testname extends Specification {
      def simpleCheck() {
          when:
          println('running simple check')
          then:
          assert false
      }
  }
  " > "$filepath"
}

function runTest() {
  echo "$separator"
  echoAndDo ./gradlew -Dorg.gradle.ignoreBuildJavaVersionCheck=true :execution:quickTest
}

testdir="platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/history/changes"

testfile1="${testdir}/${testname}-old.groovy"
testfile2="${testdir}/${testname}.groovy"

writeTest "$testname" "$testfile1"
if runTest; then
  echo "Failing test did not cause a build failure. That's weird"
  exit 1
fi
echo "$separator"
echo "Searching for the test class:"
echoAndDo find platforms/core-execution/execution/build -name "$testname*"
echo "$separator"
echo "Deleting old test"
echoAndDo rm "$testfile1"

writeTest "$testname" "$testfile2"
if runTest; then
  echo "$separator"
  echo "Notice that renaming the test file caused the build failure to go away!"
  echo "$separator"
  echo "Searching for the test class:"
  echoAndDo find platforms/core-execution/execution/build -name "$testname*"
  echo "Is the test missing?"
  exit 1
else
  echo "$separator"
  echo "Failed to reproduce the problem. Nice!"
fi
```

When I run this script inside a checkout of Gradle at tag `v8.7.0`, I get this output:

```
$ /usr/local/google/workspace/util/gradletest.sh
************************************************************************
Writing failing test at platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/history/changes/SampleTest-old.groovy
************************************************************************
./gradlew -Dorg.gradle.ignoreBuildJavaVersionCheck=true :execution:quickTest
Reusing configuration cache.

> Task :execution:compileTestGroovy
Groovy compilation avoidance is an incubating feature.

> Task :execution:test

SampleTest > simpleCheck FAILED
    org.spockframework.runtime.ConditionNotSatisfiedError at SampleTest-old.groovy:11

273 tests completed, 1 failed

> Task :execution:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':execution:test'.
> There were failing tests. See the report at: file:///usr/local/google/workspace/gradle/platforms/core-execution/execution/build/reports/tests/test/index.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.7-rc-4/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 6s
420 actionable tasks: 7 executed, 413 up-to-date
Configuration cache entry reused.
************************************************************************
Searching for the test class:
find platforms/core-execution/execution/build -name SampleTest*
platforms/core-execution/execution/build/classes/groovy/test/org/gradle/internal/execution/history/changes/SampleTest.class
************************************************************************
Deleting old test
rm platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/history/changes/SampleTest-old.groovy
************************************************************************
Writing failing test at platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/history/changes/SampleTest.groovy
************************************************************************
./gradlew -Dorg.gradle.ignoreBuildJavaVersionCheck=true :execution:quickTest
Reusing configuration cache.

BUILD SUCCESSFUL in 897ms
420 actionable tasks: 5 executed, 2 from cache, 413 up-to-date
Configuration cache entry reused.
************************************************************************
Notice that renaming the test file caused the build failure to go away!
************************************************************************
Searching for the test class:
find platforms/core-execution/execution/build -name SampleTest*
Is the test missing?
```

### Gradle version

v8.7.0

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/platforms/jvm/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy
+++ b/platforms/jvm/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy
@@ -17,10 +17,10 @@
 package org.gradle.java.compile.incremental
 
 import org.gradle.integtests.fixtures.CompiledLanguage
-import org.gradle.util.internal.ToBeImplemented
+
 import spock.lang.Issue
 
-import static org.junit.Assume.assumeTrue
+
 
 abstract class AbstractSourceIncrementalCompilationIntegrationTest extends AbstractJavaGroovyIncrementalCompilationSupport {
 
@@ -520,13 +520,7 @@ sourceSets {
     }
 
     @Issue("https://github.com/gradle/gradle/issues/28916")
-    @ToBeImplemented
     def "recompiles classes on file typo rename"() {
-        // TODO: Delete this assume statement when fixed,
-        //  since for Java Cli mode it already works accidentally,
-        //  since Cli mode doesn't track file to class mapping accurately
-        assumeTrue(this.class != JavaSourceCliIncrementalCompilationIntegrationTest.class)
-
         def fileWithTypo = file("src/main/${languageName}/ATypo.${languageName}") << """
             class A {}
         """
@@ -540,10 +534,6 @@ sourceSets {
         run language.compileTaskName
 
         then:
-        // TODO: Fix this, A should be recompiled, e.g.:
-        //   outputs.hasFiles(file("A.class"))
-        //   outputs.recompiled("A")
-        outputs.deletedClasses("A")
-        outputs.noneRecompiled()
+        outputs.recompiledClasses("A")
     }
 }

--- a/platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/AbstractRecompilationSpecProvider.java
+++ b/platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/AbstractRecompilationSpecProvider.java
@@ -28,6 +28,7 @@ import org.gradle.api.internal.tasks.compile.incremental.compilerapi.deps.Genera
 import org.gradle.api.internal.tasks.compile.incremental.transaction.CompileTransaction;
 import org.gradle.api.tasks.util.PatternSet;
 import org.gradle.internal.file.Deleter;
+import org.gradle.work.ChangeType;
 import org.gradle.work.FileChange;
 
 import java.io.File;
@@ -116,6 +117,11 @@ abstract class AbstractRecompilationSpecProvider implements RecompilationSpecPro
                 spec.setFullRebuildCause(rebuildClauseForChangedNonSourceFile(fileChange));
             }
             sourceFileChangeProcessor.processChange(changedClasses, spec);
+            // For added/modified files, record the source path directly so it's included
+            // in compilation even if the class-to-source reverse mapping is stale (see #28916).
+            if (fileChange.getChangeType() != ChangeType.REMOVED && !changedClasses.isEmpty()) {
+                spec.addSourcePath(relativeFilePath);
+            }
         }
     }
 

